File: app/src/lib/regulatory/citation-anchor.ts
import type { CitationAnchor, RegulatorySource } from "./types";

export interface CitationAnchorInput {
  source: RegulatorySource;
  section?: string;
  paragraph?: string;
  tableLabel?: string;
  pageNumber?: number;
}

export function buildCitationAnchor(input: CitationAnchorInput): CitationAnchor {
  const citationParts = [input.source.citation, input.section, input.paragraph, input.tableLabel].filter(Boolean);
  return {
    sourceId: input.source.sourceId,
    citation: citationParts.join(" "),
    section: input.section,
    paragraph: input.paragraph,
    tableLabel: input.tableLabel,
    pageNumber: input.pageNumber,
    url: input.source.url,
    retrievedAt: input.source.retrievedAt,
    sourceHash: input.source.textHash
  };
}

export function anchorToDisplayCitation(anchor: CitationAnchor) {
  return [anchor.citation, anchor.pageNumber ? `p. ${anchor.pageNumber}` : undefined].filter(Boolean).join(", ");
}

File: ingestion/traceready_ingestion/chunking/citation_anchor.py
from __future__ import annotations

from .._compat import BaseModel


class CitationAnchor(BaseModel):
    source_id: str
    citation: str
    section: str | None = None
    paragraph: str | None = None
    table_label: str | None = None
    page_number: int | None = None
    url: str | None = None
    retrieved_at: str | None = None
    source_hash: str | None = None


def build_citation_anchor(
    *,
    source_id: str,
    source_url: str,
    source_hash: str,
    retrieved_at: str,
    section: str,
    paragraph: str | None = None,
    table_label: str | None = None,
    page_number: int | None = None,
) -> CitationAnchor:
    citation = " ".join(part for part in [section, paragraph, table_label] if part)
    return CitationAnchor(
        source_id=source_id,
        citation=citation,
        section=section,
        paragraph=paragraph,
        table_label=table_label,
        page_number=page_number,
        url=source_url,
        retrieved_at=retrieved_at,
        source_hash=source_hash,
    )
